File: lab02/threaded_tree.py
class Node:
    def __init__(self, value):
        self.value = value
        self.left_child = None
        self.right_child = None
        self.left_thread = None
        self.right_thread = None


class ThreadedTree:
    def __init__(self):
        self.root = Node(-1)

    def _add_node(self, current_node, new_node):
        if new_node.value < current_node.value:
            if current_node.left_child is not None:
                current_node = current_node.left_child
                self._add_node(current_node, new_node)
            else:
                new_node.left_thread = current_node.left_thread
                new_node.right_thread = current_node
                current_node.left_child = new_node
                current_node.left_thread = None
        else:
            if current_node.right_child is not None:
                current_node = current_node.right_child
                self._add_node(current_node, new_node)
            else:
                new_node.left_thread = current_node
                new_node.right_thread = current_node.right_thread
                current_node.right_child = new_node
                current_node.rightThread = None

    def inorder_with_threads(self):
        current_node = self.root
        via_branch = True

        lst = []

        while current_node is not None:
            if via_branch:
                while current_node.left_child is not None:
                    current_node = current_node.left_child

            if current_node.value != -1: lst.append(current_node.value)

            if current_node.right_child is not None:
                via_branch = True
                current_node = current_node.right_child
            else:
                via_branch = False
                current_node = current_node.right_thread                

        return lst

    def backorder_with_threads(self):
        current_node = self.root
        via_branch = True

        lst = []

        while current_node is not None:
            if via_branch:
                while current_node.right_child is not None:
                    current_node = current_node.right_child

            if current_node.value != -1: lst.append(current_node.value)

            if current_node.left_child is not None:
                via_branch = True
                current_node = current_node.left_child
            else:
                via_branch = False
                current_node = current_node.left_thread                

        return lst

    def add_node(self, value):
        new_node = Node(value)
        self._add_node(self.root, new_node)



if __name__ == '__main__':
    lst = list(map(int, input().split()))
    threaded_tree = ThreadedTree()

    for item in lst:
        threaded_tree.add_node(item)

    print(threaded_tree.inorder_with_threads());
    print(threaded_tree.backorder_with_threads());

File: lab02/ThreadedTree.java
class Node {    
    int value;
    Node leftChild,  rightChild;
    Node leftThread, rightThread;

    Node(int val) {
        value = val;
    }
}

class OrderedTree {
    Node root;

    OrderedTree() {
        root = new Node(-1);
    }

    private void _addNode(Node currentNode, int valueToInsert) {

        if (valueToInsert < currentNode.value) {
            if (currentNode.leftChild == null) {
                Node childNode = new Node(valueToInsert);
                childNode.leftThread = currentNode.leftThread;
                childNode.rightThread = currentNode;
                currentNode.leftChild = childNode;
                currentNode.leftThread = null;
            }
            else {
                _addNode(currentNode.leftChild, valueToInsert);
            }
        }
        else {
            if (currentNode.rightChild == null) {
                Node childNode = new Node(valueToInsert);
                childNode.leftThread = currentNode;
                childNode.rightThread = currentNode.rightThread;
                currentNode.rightChild = childNode;
                currentNode.rightThread = null;
            }
            else {
                _addNode(currentNode.rightChild, valueToInsert);
            }
        }
    }

    private void _traverseInorder(Node currentNode) {
        if (currentNode.leftChild != null) {
            _traverseInorder(currentNode.leftChild);
        }

        if (currentNode.value != -1) {
            System.out.print(currentNode.value + " ");
        }

        if (currentNode.rightChild != null) {
            _traverseInorder(currentNode.rightChild);
        }
    }

    private void _traverseBackorder(Node currentNode) {
        if (currentNode.rightChild != null) {
            _traverseBackorder(currentNode.rightChild);
        }

        if (currentNode.value != -1) {
            System.out.print(currentNode.value + " ");
        }

        if (currentNode.leftChild != null) {
            _traverseBackorder(currentNode.leftChild);
        }

    }

    void inorderWithTreads() {
        Node currentNode = root;
        boolean via_branch = true;

        while (currentNode != null) {
            if (via_branch) {
                while (currentNode.leftChild != null) {
                    currentNode = currentNode.leftChild;
                }
            }

            if (currentNode.value != -1) {
                System.out.print(currentNode.value + " ");
            }

            if (currentNode.rightChild != null) {
                via_branch = true;
                currentNode = currentNode.rightChild;
            }
            else {
                via_branch = false;
                currentNode = currentNode.rightThread;
            }
        }
    }

    void backorderWithTreads() {
        Node currentNode = root;
        boolean via_branch = true;

        while (currentNode != null) {
            if (via_branch) {
                while (currentNode.rightChild != null) {
                    currentNode = currentNode.rightChild;
                }
            }

            if (currentNode.value != -1) {
                System.out.print(currentNode.value + " ");
            }

            if (currentNode.leftChild != null) {
                via_branch = true;
                currentNode = currentNode.leftChild;
            }
            else {
                via_branch = false;
                currentNode = currentNode.leftThread;
            }
        }
    }

    void traverseInorder() {
        _traverseInorder(root);
    }

    void addNode(int value) {
        _addNode(root, value);
    }

    void traverseBackorder() {
        _traverseBackorder(root);
    }
}

class Demonstration {
    public static void main(String[] args) {
        OrderedTree orderedTree = new OrderedTree();

        orderedTree.addNode(5);
        orderedTree.addNode(2);
        orderedTree.addNode(6);
        orderedTree.addNode(4);
        orderedTree.addNode(1);
        orderedTree.addNode(9);
        orderedTree.addNode(7);

        orderedTree.traverseInorder();
        System.out.println();

        orderedTree.traverseBackorder();
        System.out.println();

        orderedTree.inorderWithTreads();
        System.out.println();

        orderedTree.backorderWithTreads();
        System.out.println();

    }
}